Exporter: add SpacetimeInterpolator

nilsvu opened the pull request:
## Proposed changes

Adds the `SpacetimeInterpolator` class that can load multiple slices of volume data and then interpolate to arbitrary points in both space and time within this time range. The time range can be advanced sequentially. This is useful for ray tracing.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Exporter: add SpacetimeInterpolator

wthrowe commented on src/IO/Exporter/SpacetimeInterpolator.hpp:
```diff
@@ -0,0 +1,134 @@
+ * interpolation in space, the PointwiseInterpolator class is used, which does
+ * spectral interpolation within the elements. For the interpolation in time, a
+ * polynomial interpolation with fixed order is used (see
+ * `time_interpolation_order_`).
```
Documentation shouldn't require people to look at private member variables to find out what's going on.

wthrowe commented on src/IO/Exporter/SpacetimeInterpolator.hpp:
```diff
@@ -0,0 +1,134 @@
+  /// Advance the time bounds by loading new volume data slices. The new time
+  /// bounds will exactly follow the old time bounds (no overlap in time). This
+  /// function is not generally thread safe because it loads new data into
+  /// memory.
```
Is it unsafe to call this in parallel with other member functions, or just to call this function from multiple threads?

nilsvu replied:
This function is removed now. I added more details on thread safety to `load_time_bounds`.

wthrowe commented on src/IO/Exporter/SpacetimeInterpolator.hpp:
```diff
@@ -0,0 +1,134 @@
+  bool forward_in_time() const { return forward_in_time_; }
+
+  /// Advance the time bounds by loading new volume data slices. The new time
+  /// bounds will exactly follow the old time bounds (no overlap in time). This
```
This seems really bad from a performance point of view.  You have to block the entire evolution on I/O.  And, depending on what integrator you're using, the evaluations may not even be monotonic in time, in which case this won't work at all.

nilsvu replied:
I removed this function and added `load_time_bounds` instead. That allows for a sliding time window, and concurrently loading new data. I have tested this in a ray tracing simulation where I slide the window backwards in time while rendering multiple frames from a camera.

wthrowe commented on tests/Unit/IO/Exporter/Test_SpacetimeInterpolator.cpp:
```diff
@@ -0,0 +1,174 @@
+    CHECK(tmin < tmax);
+  } else {
+    CHECK(tmin > tmax);
+    std::tie(tmin, tmax) = std::pair{tmax, tmin};
```
`swap` is probably clearer.

wthrowe commented on src/IO/Exporter/SpacetimeInterpolator.cpp:
```diff
@@ -0,0 +1,206 @@
+                                       num_new_slices + 1,
+                                   static_cast<size_t>(current_obs_step_)};
+  if (obs_steps_bounds.first < 0 or obs_steps_bounds.second >= num_obs_steps_) {
+    ERROR("The requested time steps "
```
This should handle reaching the end of the data more gracefully.  Instead of erroring, it should decrease the end time to the maximum allowable value.

nilsvu replied:
Handled in `load_time_bounds()` now. Only time bounds within the range of available data is allowed.

nilsvu commented:
@wthrowe I updated this PR. I replaced `advance_time` with the more flexible `load_time_bounds`. That allows for a sliding time window and concurrent loading of new data.

nilsvu committed:
PyExporter: support observation selection

By observation ID, step, or time.

nilsvu committed:
PyExporter: bind PointwiseInterpolator, SpacetimeInterpolator

nilsvu committed:
PointwiseInterpolator: fix bug with time-independent domains

Time was always set to zero.

nilsvu committed:
Fix a bug in PointwiseInterpolator

wthrowe commented on src/IO/Exporter/SpacetimeInterpolator.cpp:
```diff
@@ -43,88 +43,118 @@ std::pair<int, size_t> get_observation_step(
+      new_interpolators{};
+  for (size_t i = 0; i < num_obs; ++i) {
+    if (i < interpolators_.size() and
+        interpolators_[i].obs_id() == obs_ids[i]) {
```
The indices in `interpolators_` and `obs_ids` don't necessarily correspond, so this can cause you to add copies of interpolators you already have.  I added a statement at the end of the function to print `interpolators_.size()` and then had the test loop over
```c++
interpolator.load_time_bounds({{1.5, 6.5}});
interpolator.load_time_bounds({{2.5, 6.5}});
```
and found that the number of interpolators grows by 8 every cycle.

nilsvu replied:
Thanks for catching this bug! I rewrote the function to use iterators, limits locking time further, and fixes this bug.

wthrowe commented on src/IO/Exporter/SpacetimeInterpolator.cpp:
```diff
@@ -43,88 +43,118 @@ std::pair<int, size_t> get_observation_step(
+        it = interpolators_.erase(it);
+        --it;
+      }
     }
```
This can be done much more efficiently since you know that both `interpolators_` and `obs_ids` are contiguous sequences of observations.  Both `find` and `erase` are O(N) operations being done in a loop.  Normally I wouldn't worry about this too much, but here this is done while holding an exclusive lock.  What you want to do is figure out what the range that needs to be deleted off each end of the vector is, and then do range-`erase` calls, which is O(N) total.  (Figuring out how much to delete can even be done without holding the lock.)

A similar thing holds for the other exclusive section, but that has to be rewritten anyway because of the bug causing multiple copies of the same data to be held.

wthrowe commented on src/IO/Exporter/SpacetimeInterpolator.cpp:
```diff
@@ -93,67 +93,109 @@ void SpacetimeInterpolator<Dim, Frame>::load_time_bounds(
-      }
+    if (num_erase_front > 0) {
+      interpolators_.erase(interpolators_.begin(),
+                           interpolators_.begin() + num_erase_front);
```
This can go past `end` if the new and old valid ranges are disjoint and far enough apart.  I think an easy fix is to first check something like
```c++
if (num_erase_front >= interpolators_.size() or num_erase_back >= interpoaltors_.size()) {
  interpolators_.clear();
  // copy the "interpolators_.empty()" block from above
}
```

wthrowe reviewed (changes requested):
This looks much better.  One last bug commented below, but should be an easy fix.  You can squash, including the new fix as long as it's not any more complicated than it looks from my comment.

nilsvu commented:
Done and squashed. Thanks for reviewing @wthrowe !
